# PROYECTO LÓGICA: Katas de Python

1. Escribe una función que reciba una cadena de texto como parámetro y devuelva un diccionario con las frecuencias
de cada letra en la cadena. Los espacios no deben ser considerados.

2. Dada una lista de números, obtén una nueva lista con el doble de cada valor. Usa la función map.

In [66]:
lista_números1 = [1,2,3,4,5,6,7,8,9,10]
lista_números2 = list (map (lambda x:x*2,lista_números1))
print (lista_números2)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


3. Escribe una función que tome una lista de palabras y una palabra objetivo como parámetros. La función debe
devolver una lista con todas las palabras de la lista original que contengan la palabra objetivo.

In [67]:
lista_palabras = ["manzana","pera","platano","sandia","fresas","mandarina","naranja"]
palabra_objetivo = "na"

lista_final =list(filter(palabra_objetivo, lista_palabras))
print (lista_final)

TypeError: 'str' object is not callable

4. Genera una función que calcule la diferencia entre los valores de dos listas. Usa la función map()

In [ ]:
lista1 = [1,2,3,4,5,6,7,8,9,10]
lista2 = [10,9,8,7,6,5,4,3,2,1]

diferencia = list (map (lambda x,y: x-y, lista1, lista2))
print (diferencia)

[-9, -7, -5, -3, -1, 1, 3, 5, 7, 9]


5.Ecribe una función que tome una lista de números como parámetro y un valor opcional nota_aprobado, que por
defecto es 5. La función debe calcular la media de los números en la lista y determinar si la media es mayor o igual
que nota aprobado. Si es así, el estado será "aprobado", de lo contrario, será "suspenso". La función debe devolver
una tupla que contenga la media y el estado.

In [ ]:
def calcular_media (iterable_numeros,nota_aprobado):
    estado = "Suspenso"
    media = round (sum (iterable_numeros)/len(iterable_numeros),2)
    if media >= nota_aprobado:
        estado ="Aprobado"
    return media,estado

lista_notas = [5.6,3.8,5,6.7,2.3,8,8.4]
media_final,estado_final = calcular_media(iterable_numeros=lista_notas, nota_aprobado=5)
print (f"La nota media final es: {media_final}")
print (f"El estado final es: {estado_final}")


La nota media final es: 5.69
El estado final es: Aprobado


6. Escribe una función que calcule el factorial de un número de manera recursiva.

In [119]:
def calcular_factorial (numero):
    if numero == 0:
        return 1
    else:
        return numero * calcular_factorial (numero-1)

numero_factorial = calcular_factorial(5)
print (numero_factorial)

120


7. Genera una función que convierta una lista de tuplas a una lista de strings. Usa la función map()

In [215]:
def convertir_lista_string (lista_tupla):
    resultado_conversión = list (lista_tupla)
    return resultado_conversión 

lista_tupla = (1,2,3,4,"gorka")
convertir_lista_string (lista_tupla)


[1, 2, 3, 4, 'gorka']

In [218]:
def conversion (lista1):
    return list(lista1)

lista2 = list( map (conversion, lista1))
lista1 = (1,2,3,"gorka")
print (lista2)

TypeError: 'int' object is not iterable

8. Escribe un programa que pida al usuario dos números e intente dividirlos. Si el usuario ingresa un valor no numérico o intenta dividir por cero, maneja esas excepciones de manera adecuada. Asegúrate de mostrar un mensaje
indicando si la división fue exitosa o no.

In [200]:
def dividir_numeros():
    try:
        # Solicitar al usuario el primer número
        num1 = float(input("Introduce el primer número: "))
        # Solicitar al usuario el segundo número
        num2 = float(input("Introduce el segundo número: "))
        
        # Intentar realizar la división
        resultado = num1 / num2
    except ValueError:
        # Manejar el caso en que el usuario ingrese un valor no numérico
        print("Error: Por favor, introduce valores numéricos.")
    except ZeroDivisionError:
        # Manejar el caso en que el usuario intente dividir por cero
        print("Error: No se puede dividir por cero.")
    else:
        # Si no hubo excepciones, mostrar el resultado de la división
        print(f"La división fue exitosa. El resultado es: {resultado}")

# Llamar a la función para ejecutar el programa
dividir_numeros()


Error: Por favor, introduce valores numéricos.


9. -Escribe una función que tome una lista de nombres de mascotas como parámetro y devuelva una nueva lista
excluyendo ciertas mascotas prohibidas en España. La lista de mascotas a excluir es ["Mapache", "Tigre",
"Serpiente Pitón", "Cocodrilo", "Oso"].Usa la función filter()

In [222]:
def mascotas_ok (lista_mascotas):
    if lista_mascotas == "Mapache":
        lista_mascotas_aceptadas_españa.append (lista_mascotas)
    else:
        pass
    return lista_mascotas_aceptadas_españa

lista_mascotas = ["Gato", "Perro", "Mapache","Periquito"]
mascotas_ok (lista_mascotas)

NameError: name 'lista_mascotas_aceptadas_españa' is not defined

10. Escribe una función que reciba una lista de números y calcule su promedio. Si la lista está vacía, lanza una
excepción personalizada y maneja el error adecuadamente.

In [84]:
def calcular_media (iterable_numeros):
    if len (lista_numeros)> 0:
        media = round (sum (iterable_numeros)/len(iterable_numeros),2)
        return (media)
    else:
        print ("No hay ningún valor para poder hacer la media")

lista_numeros = []
resultado_media = calcular_media (iterable_numeros = lista_numeros)
print (f"La media final es: {resultado_media}")


No hay ningún valor para poder hacer la media
La media final es: None


11. Escribe un programa que pida al usuario que introduzca su edad. Si el usuario ingresa un valor no numérico o un
valor fuera del rango esperado (por ejemplo, menor que 0 o mayor que 120), maneja las excepciones
adecuadamente.

12. Genera una función que al recibir una frase devuelva una lista con la longitud de cada palabra. Usa la función map()

In [108]:
frase = ("Buenos dias")
longitud_palabra =list (map(lambda x: len(x), frase))
print (longitud_palabra)
print (frase)

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
Buenos dias


13. Genera una función la cual, para un conjunto de caracteres, devuelva una lista de tuplas con cada letra en
mayúsculas y minúsculas. Las letras no pueden estar repetidas .Usa la función map()

14. Crea una función que retorne las palabras de una lista de palabras que comience con una letra en especifico. Usa la función filter()

15. Crea una función lambda que sume 3 a cada número de una lista dada.

In [72]:
lista_numeros = (1,2,3,4,5,6,7,8,9)
resultado = list (map(lambda x: x+3, lista_numeros))

print (resultado)

[4, 5, 6, 7, 8, 9, 10, 11, 12]


16. Escribe una función que tome una cadena de texto y un número entero n como parámetros y devuelva una lista de
todas las palabras que sean más largas que n. Usa la función filter().

In [156]:
def palabras_mas_largas_que(texto, n):
    # Dividir el texto en palabras
    palabras = texto.split()
    
    # Usar filter para obtener las palabras más largas que n
    palabras_largas = list(filter(lambda palabra: len(palabra) > n, palabras))
    
    return palabras_largas

# Ejemplo de uso
texto = "Hola mundo hoy va a ser un dia espectacular"
n = 2
resultado = palabras_mas_largas_que(texto, n)
print(resultado)  

['Hola', 'mundo', 'hoy', 'ser', 'dia', 'espectacular']


17. Crea una función que tome una lista de dígitos y devuelva el número correspondiente. Por ejemplo, [5,7,2]
corresponde al número quinientos setenta y dos (572). Usa la función reduce().

18. Escribe un programa en Python que cree una lista de diccionarios que contenga información de estudiantes
(nombre, edad, calificación) y use la función filter para extraer a los estudiantes con una calificación mayor o igual a 90. Usa la función filter() 

19. Crea una función lambda que filtre los números impares de una lista dada.

In [88]:
lista_numeros =[12,24,92,74,22,2,48,92]
numeros_impares = list (filter (lambda x: x%2 ==0, lista_numeros))
print (numeros_impares)

[12, 24, 92, 74, 22, 2, 48, 92]


20. Para una lista con elementos tipo integer y string obtén una nueva lista sólo con los valores int. Usa la función
filter() 

21. Crea una función que calcule el cubo de un número dado mediante una función lambda.


In [167]:
numero = 7
print((lambda x: x*x*x)(numero))


343


22. Dada una lista numérica, obtén el producto total de los valores de dicha lista.Usa la función reduce() .


23. Concatena una lista de palabras.Usa la función reduce() .


24. Calcula la diferencia total en los valores de una lista. Usa la función reduce() .


25. Crea una función que cuente el número de caracteres en una cadena de texto dada.

In [151]:
def contar_caracteres (cadena):
    return len(cadena)

cadena = "Hola mundo hoy va a ser un gran día"
numero_caracteres = contar_caracteres (cadena)
print (numero_caracteres)


35


26. Crea una función lambda que calcule el resto de la división entre dos números dados.


27. Crea una función que calcule el promedio de una lista de números.


In [144]:
def calcular_promedio (numeros):
    promedio = round (sum (numeros)/len(numeros),2)
    return promedio

numeros = [1,2,3,4,5,6,7,8,23,34,32,26,87,54]
promedio_numeros = calcular_promedio (numeros)
print (promedio_numeros)

20.86


28. Crea una función que busque y devuelva el primer elemento duplicado en una lista dada.
.

29. Crea una función que convierta una variable en una cadena de texto y enmascare todos los caracteres con el
carácter '#', excepto los últimos cuatro.


30. Crea una función que determine si dos palabras son anagramas, es decir, si están formadas por las mismas letras
pero en diferente orden

33. Crea una función lambda que sume elementos correspondientes de dos listas dadas.

In [150]:
lista_elementos1 = [1,2,4,65,34,38,26,6,45,93,2]
lista_elementos2 = [2,34,23,32,15,35,2]

lista_total = list (map (lambda x,y: x+y,lista_elementos1,lista_elementos2))
print (lista_total)

[3, 36, 27, 97, 49, 73, 28]
